sql: add descriptor and test catalog support for GENERATED AS IDENTITY columns

ZhouXing19 opened the pull request:
This commit adds an enum attribute to note how the column creation is
associated with the `GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY` token.
We also added corresponding test catalog support.

We implement `GeneratedAsIdentityType` enum following this logic:

1. Persistent column descriptor: Add `enum GeneratedAsIdentityType`,
and add `GeneratedAsIdentityType generated_as_id_type`
under `ColumnDescriptor`
at `pkg/sql/catalog/descpb/structured.proto`

2. Write access to `descpb.ColumnDescriptor.GeneratedAsIdType`:
set the variable when `GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY`
is in the `CREATE TABLE` statement.
Check `pkg/sql/catalog/tabledesc/table.go`

3. Read access to `descpb.ColumnDescriptor.GeneratedAsIdType`:
refactored `cat.Column.Init()` at `pkg/sql/opt/cat/column.go`

Test catalog support is added at
`pkg/sql/opt/testutils/testcat/testdata/table`.

Logic test for descriptor support:
`pkg/sql/logictest/testdata/logic_test/alter_table`
and
`pkg/sql/logictest/testdata/logic_test/create_table`.

This matches the PostgreSQL syntax:
https://www.postgresql.org/docs/current/sql-createtable.html

Release justification: although this supports a new feature, the changes are high-benefit and
have been under scrutiny for many weeks now, so after discussing with other leads we feel fine
about merging this during stability period.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68713)
<!-- Reviewable:end -->

mgartner reviewed (changes requested):
> Reviewed 16 of 16 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ZhouXing19)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhKhQXF2hjHZ68iDGXQ:-MhKhQXF2hjHZ68iDGXR:b-ffr7z8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c35c12a6dcd854c05f7df6372318310dc4cd00d/-- commits#L2)):*
This commit also updates the `catalog` package, so I'd label it as `sql: add descriptor and test catalog support for GENERATED AS IDENTITY columns`.
___
*[pkg/sql/catalog/tabledesc/validate.go, line 711 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhKlYfY22eo-GGi9jNu:-MhKlYfZ7g7NsAESo_ke:b9lyxxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c35c12a6dcd854c05f7df6372318310dc4cd00d/pkg/sql/catalog/tabledesc/validate.go#L711)):*
> ```Go
> 
> 		if column.IsNullable() && column.IsGeneratedAsIdentity() {
> 			return errors.Newf("conflicting NULL/NOT NULL declarations for column %q", column.GetName())
> ```

You should add some test cases for these errors here: https://github.com/cockroachdb/cockroach/blob/6e3438f6055730d80cb430283a3e85f7503c9e18/pkg/sql/catalog/tabledesc/validate_test.go#L312
___
*[pkg/sql/catalog/tabledesc/validate.go, line 712 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhKiKFf0kbKinpq2XlD:-MhKiKFg4RMUJumgVORo:ba6n3w0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c35c12a6dcd854c05f7df6372318310dc4cd00d/pkg/sql/catalog/tabledesc/validate.go#L712)):*
> ```Go
> 		if column.IsNullable() && column.IsGeneratedAsIdentity() {
> 			return errors.Newf("conflicting NULL/NOT NULL declarations for column %q", column.GetName())
> 		}
> ```

We should also have a check that a column is not both `GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY` and has a `DEFAULT`. This is not allowed in Postgres:

```
marcus=# CREATE TABLE t (a INT GENERATED ALWAYS AS IDENTITY DEFAULT 1);
ERROR:  42601: both default and identity specified for column "a" of table "t"
LINE 1: ...REATE TABLE t (a INT GENERATED ALWAYS AS IDENTITY DEFAULT 1)...
```

We should probably have parser tests for this case too, similar to the tests with a column that is `GENERATED ... AS IDENTITY` and has a computed expression.
___
*[pkg/sql/opt/cat/column.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhKjc2x3cUPObGcx6ne:-MhKjc2x3cUPObGcx6nf:b7u858e) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c35c12a6dcd854c05f7df6372318310dc4cd00d/pkg/sql/opt/cat/column.go#L39)):*
> ```Go
> 	computedExpr                string
> 	invertedSourceColumnOrdinal int
> 	generatedAsIdentityType     descpb.GeneratedAsIdentityType
> ```

A question for @rytaft and @raduberinde: Do we typically try to avoid types from packages like `descpb` from "leaking" into the catalog? Or is this fine?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @ZhouXing19)

___
*[pkg/sql/catalog/tabledesc/validate.go, line 712 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhKiKFf0kbKinpq2XlD:-MhLU3Wz3HcE9pJVXHE5:b4dicyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c35c12a6dcd854c05f7df6372318310dc4cd00d/pkg/sql/catalog/tabledesc/validate.go#L712)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

We should also have a check that a column is not both `GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY` and has a `DEFAULT`. This is not allowed in Postgres:

```
marcus=# CREATE TABLE t (a INT GENERATED ALWAYS AS IDENTITY DEFAULT 1);
ERROR:  42601: both default and identity specified for column "a" of table "t"
LINE 1: ...REATE TABLE t (a INT GENERATED ALWAYS AS IDENTITY DEFAULT 1)...
```

We should probably have parser tests for this case too, similar to the tests with a column that is `GENERATED ... AS IDENTITY` and has a computed expression.
</blockquote></details>

There were test for this [here](https://github.com/cockroachdb/cockroach/pull/68324/files#diff-dccba38be048680977583650bddede7895ab54cccce0fbccabcd66f2044402e8R2276-R2302) and [here](https://github.com/cockroachdb/cockroach/pull/68324/files#diff-fa7c476e990651baf64186090334103fb4acfd5ff9b37e6f699be44e27d61e61R493), but it looks like those tests are not in #69043 -- let's make sure we don't forget those tests.


<!-- Sent from Reviewable.io -->

ZhouXing19 committed:
sql: fix parser support for `GENERATED ... AS IDENTITY`

This commit is to disallow the following cases of column declaration with
`GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY` syntax under `CREATE TABLE`:

1. Both as an `IDENTITY` column and with a `DEFAULT` expression;
2. An `IDENTITY` column with a not-integer type.

Parser tests are added in `pkg/sql/parser/testdata/`.

This matches the PostgreSQL syntax:
https://www.postgresql.org/docs/current/sql-createtable.html

Release note: None

Release justification: although this supports a new feature, the changes are
high-benefit and have been under scrutiny for many weeks now, so after
discussing with other leads we feel fine about merging this during
stability period.

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhKhQXF2hjHZ68iDGXQ:-MhNDBoj4mLDZQu5tv8Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c35c12a6dcd854c05f7df6372318310dc4cd00d/-- commits#L2)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This commit also updates the `catalog` package, so I'd label it as `sql: add descriptor and test catalog support for GENERATED AS IDENTITY columns`.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/validate.go, line 711 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhKlYfY22eo-GGi9jNu:-MhNDEzCB21G-ih5VZZw:b-xtn740) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c35c12a6dcd854c05f7df6372318310dc4cd00d/pkg/sql/catalog/tabledesc/validate.go#L711)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

You should add some test cases for these errors here: https://github.com/cockroachdb/cockroach/blob/6e3438f6055730d80cb430283a3e85f7503c9e18/pkg/sql/catalog/tabledesc/validate_test.go#L312
</blockquote></details>

Done. Check [here](https://reviewable.io/reviews/cockroachdb/cockroach/68713/pkg/sql/catalog/tabledesc/validate_test.go#r1..r3).
___
*[pkg/sql/catalog/tabledesc/validate.go, line 712 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhKiKFf0kbKinpq2XlD:-MhNDQEb7f7v-s2Ad0gP:b-imls4p) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c35c12a6dcd854c05f7df6372318310dc4cd00d/pkg/sql/catalog/tabledesc/validate.go#L712)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

There were test for this [here](https://github.com/cockroachdb/cockroach/pull/68324/files#diff-dccba38be048680977583650bddede7895ab54cccce0fbccabcd66f2044402e8R2276-R2302) and [here](https://github.com/cockroachdb/cockroach/pull/68324/files#diff-fa7c476e990651baf64186090334103fb4acfd5ff9b37e6f699be44e27d61e61R493), but it looks like those tests are not in #69043 -- let's make sure we don't forget those tests.
</blockquote></details>

"Null" and "computed" conflicts are taken care of in [the parser steps](https://github.com/cockroachdb/cockroach/pull/68713/files#diff-01f73e0f6ec31b2bbaef6117ca07bc2a1870550909a5ec2bfe8b011387ad4f67R550-R577), and parser tests are added [here](https://github.com/cockroachdb/cockroach/pull/68711/files#diff-727d81af299318702bad5a0995f002fda99360e94d6701dc38f6fb305d3aba95R1102-R1149) and [here](https://github.com/cockroachdb/cockroach/pull/68711/files#diff-727d81af299318702bad5a0995f002fda99360e94d6701dc38f6fb305d3aba95R1103-R1149). In this case, is it still necessary to add similar tests in logic tests?
___
*[pkg/sql/opt/cat/column.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhKjc2x3cUPObGcx6ne:-MhNEW-RCQTuyutlQvj-:bnlk8mq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c35c12a6dcd854c05f7df6372318310dc4cd00d/pkg/sql/opt/cat/column.go#L39)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

A question for @rytaft and @raduberinde: Do we typically try to avoid types from packages like `descpb` from "leaking" into the catalog? Or is this fine?
</blockquote></details>

Changed it to `cat.GeneratedAsIdentityType` (an enum with the same structure as `descpb.GeneratedAsIdentityType`) to avoid using `descpb` in catalog.


<!-- Sent from Reviewable.io -->

mgartner commented:
___
*[pkg/sql/opt/cat/column.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhKjc2x3cUPObGcx6ne:-MhPN_eeCx6aKiE3EMyG:b-14x7ld) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c35c12a6dcd854c05f7df6372318310dc4cd00d/pkg/sql/opt/cat/column.go#L39)):*
Thanks. The issue I have with this is that the casts from the descpb enum to the cat type depend on the ordering of the different identity types being the same in the two `const` blocks. For example, if `cat.NotGeneratedAsIdentity` was not the first in its `const` block, then the cast would be incorrect.

> ```
> cat.GeneratedAsIdentityType(col.ColumnDesc().GeneratedAsIdType)
> ```

@raduberinde @rytaft Any opinions on this?


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
Great work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just some minor comments.

> Reviewed 19 of 19 files at r2, 16 of 16 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-:-MhPPBEoAais9CtaMXUj:b-q5utmd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, @rafiss, @rytaft, and @ZhouXing19)

___
*[pkg/sql/catalog/tabledesc/validate.go, line 712 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhKiKFf0kbKinpq2XlD:-MhPLTV-1yKz4sTqx_WL:b-3dp1w6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c35c12a6dcd854c05f7df6372318310dc4cd00d/pkg/sql/catalog/tabledesc/validate.go#L712)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

"Null" and "computed" conflicts are taken care of in [the parser steps](https://github.com/cockroachdb/cockroach/pull/68713/files#diff-01f73e0f6ec31b2bbaef6117ca07bc2a1870550909a5ec2bfe8b011387ad4f67R550-R577), and parser tests are added [here](https://github.com/cockroachdb/cockroach/pull/68711/files#diff-727d81af299318702bad5a0995f002fda99360e94d6701dc38f6fb305d3aba95R1102-R1149) and [here](https://github.com/cockroachdb/cockroach/pull/68711/files#diff-727d81af299318702bad5a0995f002fda99360e94d6701dc38f6fb305d3aba95R1103-R1149). In this case, is it still necessary to add similar tests in logic tests?
</blockquote></details>

I don't think it's really necessary to also add logic tests for these cases.
___
*[pkg/sql/logictest/testdata/logic_test/alter_table, line 1892 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPMpDYAnC9SmEqxhGl:-MhPMpDYAnC9SmEqxhGm:bi86hcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a890f60e29ec9bfdbd3964f8a79e1e070cc3ec9/pkg/sql/logictest/testdata/logic_test/alter_table#L1892)):*
> ```
> 
> statement ok
> DROP TABLE t;
> ```

nit: Do you want to add a `SHOW CREATE TABLE t` test here to verify that the column is added? You can leave a `# TODO(...): ...` above it if it displays as `b INT8 NOT NULL DEFAULT nextval('public.generated_always_t_b_seq'::REGCLASS)` instead of `b GENERATED ALWAYS AS IDENTITY`.
___
*[pkg/sql/logictest/testdata/logic_test/create_table, line 491 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPMvhS-PLyxUniFpa8:-MhPMvhS-PLyxUniFpa9:bi86hcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a890f60e29ec9bfdbd3964f8a79e1e070cc3ec9/pkg/sql/logictest/testdata/logic_test/create_table#L491)):*
> ```
> 
> statement ok
> CREATE TABLE generated_by_default_t_notnull (a INT UNIQUE, b INT NOT NULL GENERATED BY DEFAULT AS IDENTITY);
> ```

nit: Do you want to add a `SHOW CREATE TABLE t` test here to verify that the column is added? You can leave a `# TODO(...): ...` above it if it displays as `b INT8 NOT NULL DEFAULT nextval('public.generated_always_t_b_seq'::REGCLASS)` instead of `b GENERATED ALWAYS AS IDENTITY`.
___
*[pkg/sql/opt/cat/column.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhKjc2x3cUPObGcx6ne:-MhPQMwH1i7_3mThdKEf:b-j31b0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c35c12a6dcd854c05f7df6372318310dc4cd00d/pkg/sql/opt/cat/column.go#L39)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Thanks. The issue I have with this is that the casts from the descpb enum to the cat type depend on the ordering of the different identity types being the same in the two `const` blocks. For example, if `cat.NotGeneratedAsIdentity` was not the first in its `const` block, then the cast would be incorrect.

> ```
> cat.GeneratedAsIdentityType(col.ColumnDesc().GeneratedAsIdType)
> ```

@raduberinde @rytaft Any opinions on this?

</blockquote></details>

Maybe we should add a helper function in `opt_catalog.go` that uses a switch statement to perform this translation instead. I think that's a safer approach than a cast.
___
*[pkg/sql/sem/tree/create.go, line 552 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPKFJT-Xtjp6MTtU64:-MhPKFJU98U5RVhlubln:b-d8v1sx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c164fcba09ec0dbd91c92a602c7e7b3a7e769732/pkg/sql/sem/tree/create.go#L552)):*
> ```Go
> 		case *GeneratedAlwaysAsIdentity, *GeneratedByDefAsIdentity:
> 			if typRef.(*types.T).InternalType.Family != types.IntFamily {
> 				return nil, pgerror.Newf(pgcode.InvalidParameterValue, "identity column type must be smallint, integer, or bigint")
> ```

nit: maybe we should use the SQL datatype names in the error message: `identity column type must be INT, INT2, or INT4`. https://www.cockroachlabs.com/docs/stable/int.html


<!-- Sent from Reviewable.io -->

rytaft commented:
___
*[pkg/sql/opt/cat/column.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhKjc2x3cUPObGcx6ne:-MhPR8r87GSmA9lp0OfL:b-e0frkj) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c35c12a6dcd854c05f7df6372318310dc4cd00d/pkg/sql/opt/cat/column.go#L39)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Maybe we should add a helper function in `opt_catalog.go` that uses a switch statement to perform this translation instead. I think that's a safer approach than a cast.
</blockquote></details>

I agree with these comments. We do usually try to provide our own interface to avoid depending on other packages and to provide only the functionality needed by the optimizer. Thanks for checking!


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @RaduBerinde, @rafiss, and @ZhouXing19)

___
*[pkg/sql/opt_catalog.go, line 2298 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPW9Q29xIVtvWNYJtO:-MhPW9Q3EHVdm8eayovo:b-jgm7ib) ([raw file](https://github.com/cockroachdb/cockroach/blob/eadfca19c150e59e14a0d07e0c24c2f30aed539c/pkg/sql/opt_catalog.go#L2298)):*
> ```Go
> // the GeneratedAsIdType attribute for descpb.ColumnDescriptor.
> func mapGeneratedAsIdentityType(inType descpb.GeneratedAsIdentityType) cat.GeneratedAsIdentityType {
> 	mapGenAsIdType := map[descpb.GeneratedAsIdentityType]cat.GeneratedAsIdentityType{
> ```

nit: an array would be more efficient:

```go
typeMap := [...]cat.GeneratedAsIdentityType{
	descpb.GeneratedAsIdentityType_NOT_IDENTITY_COLUMN:  cat.NotGeneratedAsIdentity,
	descpb.GeneratedAsIdentityType_GENERATED_ALWAYS:     cat.GeneratedAlwaysAsIdentity,
	descpb.GeneratedAsIdentityType_GENERATED_BY_DEFAULT: cat.GeneratedByDefaultAsIdentity,
}
return typeMap[inType]
```


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @RaduBerinde, and @ZhouXing19)

___
*[pkg/sql/opt_catalog.go, line 2298 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPW9Q29xIVtvWNYJtO:-MhPYJC1D442Me_xsmNM:bpbhx1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/eadfca19c150e59e14a0d07e0c24c2f30aed539c/pkg/sql/opt_catalog.go#L2298)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: an array would be more efficient:

```go
typeMap := [...]cat.GeneratedAsIdentityType{
	descpb.GeneratedAsIdentityType_NOT_IDENTITY_COLUMN:  cat.NotGeneratedAsIdentity,
	descpb.GeneratedAsIdentityType_GENERATED_ALWAYS:     cat.GeneratedAlwaysAsIdentity,
	descpb.GeneratedAsIdentityType_GENERATED_BY_DEFAULT: cat.GeneratedByDefaultAsIdentity,
}
return typeMap[inType]
```

</blockquote></details>

Thanks! Made a similar `mapGeneratedAsIdentityType` function [here](https://github.com/cockroachdb/cockroach/pull/68713/files#diff-255e32cf909d2dc500dc07b8a58f0876ad9d90b18bce8b0cd54fb63e8d5adb83R2293-R2304).
___
*[pkg/sql/catalog/tabledesc/validate.go, line 712 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhKiKFf0kbKinpq2XlD:-MhPYSedAaVr6Olm_90-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c35c12a6dcd854c05f7df6372318310dc4cd00d/pkg/sql/catalog/tabledesc/validate.go#L712)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I don't think it's really necessary to also add logic tests for these cases.
</blockquote></details>

Done.
___
*[pkg/sql/opt/cat/column.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhKjc2x3cUPObGcx6ne:-MhPYWfS7KlhxbLw6k3s:b-qmv2il) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c35c12a6dcd854c05f7df6372318310dc4cd00d/pkg/sql/opt/cat/column.go#L39)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I agree with these comments. We do usually try to provide our own interface to avoid depending on other packages and to provide only the functionality needed by the optimizer. Thanks for checking!
</blockquote></details>

@rytaft @mgartner Yes that makes sense, added a [`mapGeneratedAsIdentityType`](https://github.com/cockroachdb/cockroach/pull/68713/files#diff-255e32cf909d2dc500dc07b8a58f0876ad9d90b18bce8b0cd54fb63e8d5adb83R2293-R2304) for this. Thanks!
___
*[pkg/sql/sem/tree/create.go, line 552 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPKFJT-Xtjp6MTtU64:-MhPXqm998ExAr3uJ2GQ:b-59gh8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/c164fcba09ec0dbd91c92a602c7e7b3a7e769732/pkg/sql/sem/tree/create.go#L552)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: maybe we should use the SQL datatype names in the error message: `identity column type must be INT, INT2, or INT4`. https://www.cockroachlabs.com/docs/stable/int.html
</blockquote></details>

Done, check [here](https://github.com/cockroachdb/cockroach/pull/68713/files#diff-01f73e0f6ec31b2bbaef6117ca07bc2a1870550909a5ec2bfe8b011387ad4f67R552).


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
___
*[pkg/sql/logictest/testdata/logic_test/alter_table, line 1892 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPMpDYAnC9SmEqxhGl:-MhPbOAd4ufXlR58-gZw:bs7wumb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a890f60e29ec9bfdbd3964f8a79e1e070cc3ec9/pkg/sql/logictest/testdata/logic_test/alter_table#L1892)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Do you want to add a `SHOW CREATE TABLE t` test here to verify that the column is added? You can leave a `# TODO(...): ...` above it if it displays as `b INT8 NOT NULL DEFAULT nextval('public.generated_always_t_b_seq'::REGCLASS)` instead of `b GENERATED ALWAYS AS IDENTITY`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/logictest/testdata/logic_test/create_table, line 491 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPMvhS-PLyxUniFpa8:-MhPbPUtAM9E8Vpziduz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a890f60e29ec9bfdbd3964f8a79e1e070cc3ec9/pkg/sql/logictest/testdata/logic_test/create_table#L491)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Do you want to add a `SHOW CREATE TABLE t` test here to verify that the column is added? You can leave a `# TODO(...): ...` above it if it displays as `b INT8 NOT NULL DEFAULT nextval('public.generated_always_t_b_seq'::REGCLASS)` instead of `b GENERATED ALWAYS AS IDENTITY`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Heads up, @postamar would like to give the catalog changes a review

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-:-MhPdiUy5hnU-pr1ipxJ:b-of7w9e)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @RaduBerinde)



<!-- Sent from Reviewable.io -->

RichardJCai reviewed (changes requested):
> Reviewed 2 of 19 files at r5, 14 of 16 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @RaduBerinde, and @ZhouXing19)

___
*[pkg/sql/catalog/table_elements.go, line 280 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPi2rL4Umy61xdWV0V:-MhPi2rL4Umy61xdWV0W:b-cy61fi) ([raw file](https://github.com/cockroachdb/cockroach/blob/2394d1c147d433d7c215183fb4a299a3628f2aba/pkg/sql/catalog/table_elements.go#L280)):*
> ```Go
> 	IsSystemColumn() bool
> 
> 	// IsGeneratedAsIdentity returns true iff the columns is created
> ```

Nit: columns/column
___
*[pkg/sql/catalog/descpb/structured.proto, line 157 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPi9Y9Dv46mnghw9Bj:-MhPi9Y9Dv46mnghw9Bk:b-5qf9lp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2394d1c147d433d7c215183fb4a299a3628f2aba/pkg/sql/catalog/descpb/structured.proto#L157)):*
> ```protobuf
>   // If the column is created with GENERATED BY DEFAULT AS IDENTITY token,
>   // the GeneratedAsIdentityType for this column will be set to GENERATED_BY_DEFAULT.
>   // If the column is created without using GENERATED ... AS IDENTITY token,
> ```

nit: without using the
___
*[pkg/sql/catalog/tabledesc/column.go, line 190 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPiGTTChNQgb0WHV0j:-MhPiGTTChNQgb0WHV0k:b-xocivy) ([raw file](https://github.com/cockroachdb/cockroach/blob/2394d1c147d433d7c215183fb4a299a3628f2aba/pkg/sql/catalog/tabledesc/column.go#L190)):*
> ```Go
> }
> 
> // IsGeneratedAsIdentity returns true iff the columns is created
> ```

nit: columns/column
___
*[pkg/sql/catalog/tabledesc/validate_test.go, line 1111 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPiZRp0DOyd97IldGw:-MhPiZRp0DOyd97IldGx:b-2qj2pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2394d1c147d433d7c215183fb4a299a3628f2aba/pkg/sql/catalog/tabledesc/validate_test.go#L1111)):*
> ```Go
> 				NextIndexID:  3,
> 			}},
> 		{`conflicting NULL/NOT NULL declarations for column "bar"`,
> ```

Should there be a test if a default expression and generated as identity is specified? Or is that disallowed in the syntax.


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
___
*[pkg/sql/catalog/tabledesc/validate_test.go, line 1111 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPiZRp0DOyd97IldGw:-MhPjydRE4arFvoBWJBh:bwqmbp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2394d1c147d433d7c215183fb4a299a3628f2aba/pkg/sql/catalog/tabledesc/validate_test.go#L1111)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Should there be a test if a default expression and generated as identity is specified? Or is that disallowed in the syntax.
</blockquote></details>

There'll be an implicit default expression (a sequence) for a `IDNETITY` column, so we take care of this case only in the parser step, check [here](https://github.com/cockroachdb/cockroach/pull/68713/files#diff-01f73e0f6ec31b2bbaef6117ca07bc2a1870550909a5ec2bfe8b011387ad4f67R537-R561). 

I.e. we only disallow a statement like  `a INT GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY DEFAULT expr` or `a INT  DEFAULT expr GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY`.


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @RaduBerinde, and @RichardJCai)

___
*[pkg/sql/catalog/table_elements.go, line 280 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPi2rL4Umy61xdWV0V:-MhPo_cI-XtX_XKJqabj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2394d1c147d433d7c215183fb4a299a3628f2aba/pkg/sql/catalog/table_elements.go#L280)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Nit: columns/column
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descpb/structured.proto, line 157 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPi9Y9Dv46mnghw9Bj:-MhPoaFE4tK-S664BYWl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2394d1c147d433d7c215183fb4a299a3628f2aba/pkg/sql/catalog/descpb/structured.proto#L157)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

nit: without using the
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/column.go, line 190 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPiGTTChNQgb0WHV0j:-MhPoYuG4mg0wqiJjDJR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2394d1c147d433d7c215183fb4a299a3628f2aba/pkg/sql/catalog/tabledesc/column.go#L190)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

nit: columns/column
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
___
*[pkg/sql/create_table.go, line 1545 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPyRPu6JdX8lzgMQ8z:-MhPyRPu6JdX8lzgMQ9-:b-g43mz7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f102759172554c3c2bce3c664ea082f59fe801d/pkg/sql/create_table.go#L1545)):*
> ```Go
> 					return nil, pgerror.Newf(pgcode.FeatureNotSupported,
> 						"version %v must be finalized to use GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY expression",
> 						clusterversion.ExpressionIndexes)
> ```

@mgartner @rafiss This version check is added per [this comment](https://github.com/cockroachdb/cockroach/pull/67956#pullrequestreview-722417038).


<!-- Sent from Reviewable.io -->

postamar commented on pkg/sql/catalog/descpb/structured.proto:
```diff
@@ -148,6 +148,17 @@ message ColumnDescriptor {
+  // If the column is created without using GENERATED ... AS IDENTITY token,
+  // GeneratedAsIdentityType for this column will be set to the default
+  // NOT_IDENTITY_COLUMN.
+  optional GeneratedAsIdentityType generated_as_id_type = 18 [(gogoproto.nullable) = false];
```
`id` is a bit overloaded and might lead to confusion. May I recommend renaming to `generated_as_ident_type` if `generated_as_identity_type` is not possible?

postamar commented on pkg/sql/catalog/tabledesc/column.go:
```diff
@@ -187,6 +187,12 @@ func (w column) IsSystemColumn() bool {
+func (w column) IsGeneratedAsIdentity() bool {
+	return w.desc.GeneratedAsIdType != descpb.GeneratedAsIdentityType_NOT_IDENTITY_COLUMN
+}
+
```
Perhaps add another method that returns `w.desc.GeneratedAsIdType`? This way you don't need to retrieve the enum value via `.ColumnDesc()` when building the opt catalog. which is something we want to discourage.

postamar replied:
In fact looking at your subsequent PRs I noticed that you added these in a subsequent commit. Perhaps move them into this commit? Your subsequent changes in `pkg/sql/catalog/schemaexpr/column.go` would also belong in this commit.

postamar commented on pkg/sql/catalog/tabledesc/validate_test.go:
```diff
@@ -1107,6 +1108,50 @@ func TestValidateTableDesc(t *testing.T) {
+					{ID: 1, Name: "bar", GeneratedAsIdType: descpb.GeneratedAsIdentityType_GENERATED_ALWAYS, ComputeExpr: &computedExpr},
+				},
+				NextColumnID: 3,
+			}},
```
Thanks for adding these test cases!

postamar reviewed (commented):
Reviewing on behalf of @cockroachdb/sql-schema I just have a few nits.

ZhouXing19 commented:
___
*[pkg/sql/catalog/descpb/structured.proto, line 160 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MgpsCyql92JWEkJ2Hgf-r10-160:-MhQEQXY1s0Pxkq2FvMk:b3t5x18) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f102759172554c3c2bce3c664ea082f59fe801d/pkg/sql/catalog/descpb/structured.proto#L160)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

`id` is a bit overloaded and might lead to confusion. May I recommend renaming to `generated_as_ident_type` if `generated_as_identity_type` is not possible?
</blockquote></details>

Thanks for reviewing! Done renaming all `Id` to `Identity`.


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @postamar, @RaduBerinde, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/catalog/tabledesc/column.go, line 195 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-Mh0FAHlZm7W8aEAJs_w-r10-195:-MhUwCLk-U1eUbDWTkbG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f102759172554c3c2bce3c664ea082f59fe801d/pkg/sql/catalog/tabledesc/column.go#L195)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

In fact looking at your subsequent PRs I noticed that you added these in a subsequent commit. Perhaps move them into this commit? Your subsequent changes in `pkg/sql/catalog/schemaexpr/column.go` would also belong in this commit.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/create_table.go:
```diff
@@ -1537,6 +1538,13 @@ func NewTableDesc(
 				d.Computed.Expr = schemaexpr.MaybeRewriteComputedColumn(d.Computed.Expr, evalCtx.SessionData)
 			}
+			if d.GeneratedIdentity.IsGeneratedAsIdentity {
+				if !evalCtx.Settings.Version.IsActive(ctx, clusterversion.ExpressionIndexes) {
```
ah, well these aren't expression indexes, so instead we should add a new clusterversion in pkg/clusterversion/cockroach_versions.go. you can make a new key here https://github.com/cockroachdb/cockroach/blob/65457f9df618177bfd56096c9be0194db2ef5cf8/pkg/clusterversion/cockroach_versions.go#L298 named `GeneratedAsIdentity` and add the version number for it here https://github.com/cockroachdb/cockroach/blob/65457f9df618177bfd56096c9be0194db2ef5cf8/pkg/clusterversion/cockroach_versions.go#L498

the idea is that once a cluster reaches that version key, it can't be downgraded back to an older version

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just a couple of nits

> Reviewed 2 of 19 files at r5, 19 of 19 files at r12, 18 of 18 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-:-MhVLrGU1C90BdyMfE8b:b-u3zw1l)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @postamar, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/opt_catalog.go, line 2298 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPW9Q29xIVtvWNYJtO:-MhVISCoDm873mOu_BZC:bbb9pek) ([raw file](https://github.com/cockroachdb/cockroach/blob/eadfca19c150e59e14a0d07e0c24c2f30aed539c/pkg/sql/opt_catalog.go#L2298)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Thanks! Made a similar `mapGeneratedAsIdentityType` function [here](https://github.com/cockroachdb/cockroach/pull/68713/files#diff-255e32cf909d2dc500dc07b8a58f0876ad9d90b18bce8b0cd54fb63e8d5adb83R2293-R2304).
</blockquote></details>

Looks like you're still using a map instead of array.
___
*[pkg/sql/catalog/tabledesc/validate_test.go, line 1151 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhVKVWr-Dp1aRE4xFR8:-MhVKVWr-Dp1aRE4xFR9:bum0oqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5574bd61a0792c69be7da94f5fc5079acdd6750e/pkg/sql/catalog/tabledesc/validate_test.go#L1151)):*
> ```Go
> 				FormatVersion: descpb.InterleavedFormatVersion,
> 				Columns: []descpb.ColumnDescriptor{
> 					{ID: 1, Name: "bar", GeneratedAsIdentityType: descpb.GeneratedAsIdentityType_GENERATED_ALWAYS, ComputeExpr: &computedExpr},
> ```

should this one be `GENERATED_BY_DEFAULT`? (otherwise this looks like a duplicate)


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
___
*[pkg/sql/create_table.go, line 1542 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPy4vlnnOnWNAX9-R7-r13-1542:-Mha5R2t6_xU6A6A8LEX:b-vjo1zc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5574bd61a0792c69be7da94f5fc5079acdd6750e/pkg/sql/create_table.go#L1542)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

ah, well these aren't expression indexes, so instead we should add a new clusterversion in pkg/clusterversion/cockroach_versions.go. you can make a new key here https://github.com/cockroachdb/cockroach/blob/65457f9df618177bfd56096c9be0194db2ef5cf8/pkg/clusterversion/cockroach_versions.go#L298 named `GeneratedAsIdentity` and add the version number for it here https://github.com/cockroachdb/cockroach/blob/65457f9df618177bfd56096c9be0194db2ef5cf8/pkg/clusterversion/cockroach_versions.go#L498

the idea is that once a cluster reaches that version key, it can't be downgraded back to an older version
</blockquote></details>

Awesome! Should I set the version key `21.1.7` or some later version?


<!-- Sent from Reviewable.io -->

ZhouXing19 committed:
sql: add descriptor and test catalog support for GENERATED AS IDENTITY columns

This commit adds an enum attribute to note how the column creation is
associated with the `GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY` token.
We also added corresponding test catalog support.

We implement `GeneratedAsIdentityType` enum following this logic:

1. Persistent column descriptor: Add `enum GeneratedAsIdentityType`,
and add `GeneratedAsIdentityType generated_as_id_type`
under `ColumnDescriptor`
at `pkg/sql/catalog/descpb/structured.proto`

2. Write access to `descpb.ColumnDescriptor.GeneratedAsIdType`:
set the variable when `GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY`
is in the `CREATE TABLE` statement.
Check `pkg/sql/catalog/tabledesc/table.go`

3. Read access to `descpb.ColumnDescriptor.GeneratedAsIdType`:
refactored `cat.Column.Init()` at `pkg/sql/opt/cat/column.go`

Test catalog support is added at
`pkg/sql/opt/testutils/testcat/testdata/table`.

Logic test for descriptor support:
`pkg/sql/logictest/testdata/logic_test/alter_table`
and
`pkg/sql/logictest/testdata/logic_test/create_table`.

This matches the PostgreSQL syntax:
https://www.postgresql.org/docs/current/sql-createtable.html

Release note: None

Release justification: although this supports a new feature,
the changes are high-benefit and have been under scrutiny for many weeks now,
so after discussing with other leads we feel fine about merging this during
stability period.

ajwerner reviewed (approved):
My sense is that somewhere we're going to want a release note. 

> Reviewed 4 of 4 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-:-MhnPU1J3MGg5AwNZfWE:bjlible)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @postamar, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[pkg/clusterversion/cockroach_versions.go, line 481 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhnO_FHD80zb7FyHiZn:-MhnO_FHD80zb7FyHiZo:b-v7fsbd) ([raw file](https://github.com/cockroachdb/cockroach/blob/285ae980240aaa30ce441374bf9639360c2946aa/pkg/clusterversion/cockroach_versions.go#L481)):*
> ```Go
> 	{
> 		Key:     GeneratedAsIdentity,
> 		Version: roachpb.Version{Major: 21, Minor: 1, Internal: 136},
> ```

this is going to need a rebase


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajwerner, @mgartner, @postamar, @rafiss, @RichardJCai, @rytaft, and @ZhouXing19)

___
*[pkg/clusterversion/cockroach_versions.go, line 481 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhnO_FHD80zb7FyHiZn:-MhnmbSNDm5Z02_g-1Ek:bs7wumb) ([raw file](https://github.com/cockroachdb/cockroach/blob/285ae980240aaa30ce441374bf9639360c2946aa/pkg/clusterversion/cockroach_versions.go#L481)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this is going to need a rebase
</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 1542 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPy4vlnnOnWNAX9-R7-r13-1542:-Mhnmkg5C7cAY5gmEMu2:bwspv52) ([raw file](https://github.com/cockroachdb/cockroach/blob/5574bd61a0792c69be7da94f5fc5079acdd6750e/pkg/sql/create_table.go#L1542)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Awesome! Should I set the version key `21.1.7` or some later version?
</blockquote></details>

Done, check [here](https://reviewable.io/reviews/cockroachdb/cockroach/68713/pkg/clusterversion/cockroach_versions.go#r14..r16).
___
*[pkg/sql/opt_catalog.go, line 2298 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhPW9Q29xIVtvWNYJtO:-MhnmpAUA19SK7sAVZJd:bk3a0n3) ([raw file](https://github.com/cockroachdb/cockroach/blob/eadfca19c150e59e14a0d07e0c24c2f30aed539c/pkg/sql/opt_catalog.go#L2298)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Looks like you're still using a map instead of array.
</blockquote></details>

Thanks! Changed it to an array.
___
*[pkg/sql/catalog/tabledesc/validate_test.go, line 1151 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhVKVWr-Dp1aRE4xFR8:-MhnmvME7glZe4kETP1E:bnugec1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5574bd61a0792c69be7da94f5fc5079acdd6750e/pkg/sql/catalog/tabledesc/validate_test.go#L1151)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

should this one be `GENERATED_BY_DEFAULT`? (otherwise this looks like a duplicate)
</blockquote></details>

Yes it's a duplicated test, have made the change. Thanks!


<!-- Sent from Reviewable.io -->

rafiss commented:
> My sense is that somewhere we're going to want a release note.

the release note will come with this PR: https://github.com/cockroachdb/cockroach/pull/69107

this looks great! feel free to merge. (I'm also adding a release justification to the PR as per https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/168656965/Guidelines+for+Release+Stability+Period

mgartner reviewed (approved):
> Reviewed 2 of 19 files at r5, 1 of 19 files at r12, 19 of 19 files at r15, 22 of 22 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @postamar, @rafiss, @RichardJCai, @rytaft, and @ZhouXing19)

___
*[pkg/sql/create_table.go, line 1544 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-Mhnt8tnE4Zsk4uKHUjo:-Mhnt8tnE4Zsk4uKHUjp:b9zlfzj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1356273f7ff617b379e2985e503032419b2a6bd7/pkg/sql/create_table.go#L1544)):*
> ```Go
> 			}
> 			if d.GeneratedIdentity.IsGeneratedAsIdentity {
> 				if !evalCtx.Settings.Version.IsActive(ctx, clusterversion.GeneratedAsIdentity) {
> ```

Does `ALTER TABLE` use this code path, or do you need to add this same check somewhere else to ensure that the correct version is finalized for `ALTER TABLE`?
___
*[pkg/sql/logictest/testdata/logic_test/alter_table, line 1889 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhnutPQ3QKfJMLbdJ-M:-MhnutPQ3QKfJMLbdJ-N:b7c19ej) ([raw file](https://github.com/cockroachdb/cockroach/blob/1356273f7ff617b379e2985e503032419b2a6bd7/pkg/sql/logictest/testdata/logic_test/alter_table#L1889)):*
> ```
> 
> statement ok
> ALTER TABLE t ADD COLUMN c INT GENERATED BY DEFAULT AS IDENTITY
> ```

nit: you could add the `SHOW CREATE` + `TODO` tests here, like the ones you added in `pkg/sql/logictest/testdata/logic_test/create_table`
___
*[pkg/sql/opt/cat/column.go, line 272 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhnvRz0-PEnCjxmy-nB:-MhnvRz0-PEnCjxmy-nC:bk6ahql) ([raw file](https://github.com/cockroachdb/cockroach/blob/1356273f7ff617b379e2985e503032419b2a6bd7/pkg/sql/opt/cat/column.go#L272)):*
> ```Go
> 		if generatedAsIdentityType != NotGeneratedAsIdentity {
> 			panic(pgerror.Newf(pgcode.Syntax,
> 				"both identity and generation expression specified for column %q", name))
> ```

this should be an assertion error: `panic(errors.AssertionFailedf("both identity ...", name))`


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
___
*[pkg/sql/opt/cat/column.go, line 272 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhnvRz0-PEnCjxmy-nB:-Mhnx-bO9BC-5UytO3AI:b-dhmij3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1356273f7ff617b379e2985e503032419b2a6bd7/pkg/sql/opt/cat/column.go#L272)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

this should be an assertion error: `panic(errors.AssertionFailedf("both identity ...", name))`
</blockquote></details>

But postgres seems to give a `syntax error` error code:

```
postgres=# \set VERBOSITY verbose
postgres=# CREATE TABLE demo_t1 (x INT GENERATED ALWAYS AS IDENTITY GENERATED ALWAYS AS (1+1) STORED);
ERROR:  42601: both identity and generation expression specified for column "x" of table "demo_t1"
LINE 1: ...TABLE demo_t1 (x INT GENERATED ALWAYS AS IDENTITY GENERATED ...
                                                             ^
LOCATION:  transformColumnDefinition, parse_utilcmd.c:837
``` 

And search 42601 in https://www.postgresql.org/docs/12/errcodes-appendix.html gives `syntax error`.


<!-- Sent from Reviewable.io -->

mgartner commented:
___
*[pkg/sql/opt/cat/column.go, line 272 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhnvRz0-PEnCjxmy-nB:-Mhnz5n85mAnGolyQXn2:b-4jfr23) ([raw file](https://github.com/cockroachdb/cockroach/blob/1356273f7ff617b379e2985e503032419b2a6bd7/pkg/sql/opt/cat/column.go#L272)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

But postgres seems to give a `syntax error` error code:

```
postgres=# \set VERBOSITY verbose
postgres=# CREATE TABLE demo_t1 (x INT GENERATED ALWAYS AS IDENTITY GENERATED ALWAYS AS (1+1) STORED);
ERROR:  42601: both identity and generation expression specified for column "x" of table "demo_t1"
LINE 1: ...TABLE demo_t1 (x INT GENERATED ALWAYS AS IDENTITY GENERATED ...
                                                             ^
LOCATION:  transformColumnDefinition, parse_utilcmd.c:837
``` 

And search 42601 in https://www.postgresql.org/docs/12/errcodes-appendix.html gives `syntax error`.
</blockquote></details>

Correct, but you've already added the syntax error for this in the parser: https://github.com/cockroachdb/cockroach/pull/68713/files#diff-01f73e0f6ec31b2bbaef6117ca07bc2a1870550909a5ec2bfe8b011387ad4f67R562-R565

Therefore, if a user tried to make a computed identity column, this code wouldn't be executed - the parser would return an error first. If somehow this check failed it should be an internal error.


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @postamar, @rafiss, @RichardJCai, @rytaft, and @ZhouXing19)

___
*[pkg/sql/create_table.go, line 1544 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-Mhnt8tnE4Zsk4uKHUjo:-MhnyAwa8EO4OqhfV_hb:bhpuo4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1356273f7ff617b379e2985e503032419b2a6bd7/pkg/sql/create_table.go#L1544)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Does `ALTER TABLE` use this code path, or do you need to add this same check somewhere else to ensure that the correct version is finalized for `ALTER TABLE`?
</blockquote></details>

`ALTER TABLE ... ADD COLUMN` should have been supported 
___
*[pkg/sql/logictest/testdata/logic_test/alter_table, line 1889 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhnutPQ3QKfJMLbdJ-M:-Mho1zTmBHFEXHGauSif:b-k2fiv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1356273f7ff617b379e2985e503032419b2a6bd7/pkg/sql/logictest/testdata/logic_test/alter_table#L1889)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: you could add the `SHOW CREATE` + `TODO` tests here, like the ones you added in `pkg/sql/logictest/testdata/logic_test/create_table`
</blockquote></details>

Yeah, added `CREATE TABLE` [here](a3aad67844), and also the corrected version is in #69107, check [here](https://github.com/cockroachdb/cockroach/pull/69107/commits/b2ed3facd7f6374e63b15fb2b243dec849b303d3#diff-9f2dd101471409a892e6774c1e8a876db2d72f8d47d9e467513d9ae0ccc27f88)
___
*[pkg/sql/opt/cat/column.go, line 272 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhnvRz0-PEnCjxmy-nB:-Mho2Rc420YhC3bqsgy8:beji3dm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1356273f7ff617b379e2985e503032419b2a6bd7/pkg/sql/opt/cat/column.go#L272)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Correct, but you've already added the syntax error for this in the parser: https://github.com/cockroachdb/cockroach/pull/68713/files#diff-01f73e0f6ec31b2bbaef6117ca07bc2a1870550909a5ec2bfe8b011387ad4f67R562-R565

Therefore, if a user tried to make a computed identity column, this code wouldn't be executed - the parser would return an error first. If somehow this check failed it should be an internal error.
</blockquote></details>

Yeah that makes sense, already made the change. Thanks!


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @postamar, @rafiss, @RichardJCai, @rytaft, and @ZhouXing19)

___
*[pkg/sql/create_table.go, line 1544 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-Mhnt8tnE4Zsk4uKHUjo:-MhoRnr-CYF10QE0IVWo:bdtlph0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1356273f7ff617b379e2985e503032419b2a6bd7/pkg/sql/create_table.go#L1544)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

`ALTER TABLE ... ADD COLUMN` should have been supported 
</blockquote></details>

@mgartner Thanks! Added the version check to `ALTER TABLE` [here](https://github.com/cockroachdb/cockroach/pull/68713/files#diff-68f42b31f1c350a76fc10b08db808133d8a2d0aa35397c246d20824c93f5e9df).


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 19 files at r15, 20 of 22 files at r16, 3 of 3 files at r17, 1 of 1 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-:-MhoSu1UCI_Y315QYLJ7:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @postamar, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[-- commits, line 52 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhoS55i-0GDKZwVvVSZ:-MhoS55i-0GDKZwVvVS_:b-at4ywr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1614c3491b28d2f00c33215ae2c861c293def3a0/-- commits#L52)):*
Looks like you'll need to fill this in
___
*[pkg/clusterversion/cockroach_versions.go, line 305 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhoS9Vb7z2xGY7dd2at:-MhoS9Vc37-aoVpes_IJ:b-m727sx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1614c3491b28d2f00c33215ae2c861c293def3a0/pkg/clusterversion/cockroach_versions.go#L305)):*
> ```Go
> 	UseKeyEncodeForHashShardedIndexes
> 	// GeneratedAsIdentity is the syntax support for `GENERATED {ALWAYS | BY
> 	//DEFAULT} AS IDENTITY` under `CREATE TABLE` syntax.
> ```

nit: space between // and GENERATED (sorry I've made this and other comments on all the PRs, not sure which commit goes with which anymore)


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r17, 1 of 1 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-:-Mhob71G0-MuNvWInOSI:bg7igkv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner, @postamar, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/opt/cat/column.go, line 269 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhoaxDUB8DXmg_UAK3R:-MhoaxDUB8DXmg_UAK3S:bi90ouo) ([raw file](https://github.com/cockroachdb/cockroach/blob/1614c3491b28d2f00c33215ae2c861c293def3a0/pkg/sql/opt/cat/column.go#L269)):*
> ```Go
> 	if computedExpr != nil {
> 		if generatedAsIdentityType != NotGeneratedAsIdentity {
> 			panic(errors.AssertionFailedf("both identity and generation expression specified for column %q", name))
> ```

nit: "both generated identity and computed expression specified for column %q"

Sorry I missed this earlier.


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @postamar, @rafiss, @RichardJCai, and @rytaft)

___
*[-- commits, line 52 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhoS55i-0GDKZwVvVSZ):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Looks like you'll need to fill this in
</blockquote></details>

Done.
___
*[pkg/clusterversion/cockroach_versions.go, line 305 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhoS9Vb7z2xGY7dd2at:-Mhq4cDS05khiIJnZ-Fk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1614c3491b28d2f00c33215ae2c861c293def3a0/pkg/clusterversion/cockroach_versions.go#L305)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: space between // and GENERATED (sorry I've made this and other comments on all the PRs, not sure which commit goes with which anymore)
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/validate_test.go, line 1154 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-Mh0HxHB8CosVHdCqkjW-r10-1154:-Mhq4mMJE_wOAo8fSrLx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f102759172554c3c2bce3c664ea082f59fe801d/pkg/sql/catalog/tabledesc/validate_test.go#L1154)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Thanks for adding these test cases!
</blockquote></details>

Done.
___
*[pkg/sql/opt/cat/column.go, line 269 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhoaxDUB8DXmg_UAK3R:-Mhq4n53AxcLPbL33e0Y:b-df25jb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1614c3491b28d2f00c33215ae2c861c293def3a0/pkg/sql/opt/cat/column.go#L269)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: "both generated identity and computed expression specified for column %q"

Sorry I missed this earlier.
</blockquote></details>

Done, changed all `both identity and generation expression specified for column` to `both generated identity and computed expression specified for column`.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (changes requested):
> Reviewed 3 of 10 files at r20, 23 of 23 files at r22, 25 of 25 files at r23, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @postamar, @rafiss, @rytaft, and @ZhouXing19)

___
*[pkg/sql/catalog/tabledesc/table.go, line 82 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhssZc18iOj14-sJmqR:-MhssZc18iOj14-sJmqS:bhybe18) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c8834b669a30676aafc2c1f97be78d85514356b/pkg/sql/catalog/tabledesc/table.go#L82)):*
> ```Go
> 		case tree.GeneratedByDefault:
> 			col.GeneratedAsIdentityType = descpb.GeneratedAsIdentityType_GENERATED_BY_DEFAULT
> 		}
> ```

nit: add default in switch and return assertion error
___
*[pkg/sql/opt/testutils/testcat/create_table.go, line 613 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-Mhst-o30BObJZmqz6xh:-Mhst-o30BObJZmqz6xi:bhh7qha) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c8834b669a30676aafc2c1f97be78d85514356b/pkg/sql/opt/testutils/testcat/create_table.go#L613)):*
> ```Go
> 		default:
> 			panic(fmt.Errorf(
> 				"column %s is of wrong generated as identity type (neither ALWAYS nor BY DEFAULT)",
> ```

assertion error instead of errorf


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (approved):
LGTM modulo error nits

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-:-Mhst6J38Zt0lEIO64x_:b-e742ej)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @postamar, @rafiss, @rytaft, and @ZhouXing19)



<!-- Sent from Reviewable.io -->

mgartner commented:
___
*[pkg/sql/opt/testutils/testcat/create_table.go, line 613 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-Mhst-o30BObJZmqz6xh:-Mhsy_IR9kYyq8BsZAug:b-ndedg3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c8834b669a30676aafc2c1f97be78d85514356b/pkg/sql/opt/testutils/testcat/create_table.go#L613)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

assertion error instead of errorf
</blockquote></details>

This is fine as an errorf because this is the test catalog. This code only runs during optimizer tests. You can see other examples of using errorf in this file.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :shipit: 

> Reviewed 3 of 10 files at r20, 25 of 25 files at r23, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-:-MhsysMP1nzxPvQG4_-z:byzm7aw)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @postamar, @rafiss, @RichardJCai, @rytaft, and @ZhouXing19)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
looks good! feel free to merge :)

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @postamar, @rafiss, @RichardJCai, and @rytaft)

___
*[pkg/sql/catalog/tabledesc/table.go, line 82 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-MhssZc18iOj14-sJmqR:-Mht1HGvDMYP5_ivyB8C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c8834b669a30676aafc2c1f97be78d85514356b/pkg/sql/catalog/tabledesc/table.go#L82)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

nit: add default in switch and return assertion error
</blockquote></details>

Done.
___
*[pkg/sql/opt/testutils/testcat/create_table.go, line 613 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/68713#-Mhst-o30BObJZmqz6xh:-Mht1VAeCKEHaat4zXlw:b6zhzfh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c8834b669a30676aafc2c1f97be78d85514356b/pkg/sql/opt/testutils/testcat/create_table.go#L613)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This is fine as an errorf because this is the test catalog. This code only runs during optimizer tests. You can see other examples of using errorf in this file.
</blockquote></details>

Thanks! I'll keep with a errorf.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 3 of 10 files at r20, 4 of 23 files at r21, 23 of 25 files at r23, 2 of 2 files at r25, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @postamar, @rafiss, and @RichardJCai)



<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
> Reviewed 2 of 2 files at r25, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68713)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @postamar, @rafiss, and @RichardJCai)



<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
bors r+

ZhouXing19 commented:
bors r+

ZhouXing19 commented:
bors r+
